File: static/js/layout-convert.js
export let gLyt = { lbl: "q" };
const layout_string = {
  'qwerty.low'    	: String.raw`${"`"}1234567890-=`     +`qwertyuiop[]`+`asdfghjkl;'`+`zxcvbnm,./`,
  'qwerty.upp'    	: String.raw     `~!@#$%^&*()_+`     +`QWERTYUIOP{}`+`ASDFGHJKL:"`+`ZXCVBNM<>?`,
  'dvorak.low'    	: String.raw`${"`"}1234567890[]`     +`',.pyfgcrl/=`+`aoeuidhtns-`+`;qjkxbmwvz`,
  'dvorak.upp'    	: String.raw     `~!@#$%^&*(){}`     +`"<>PYFGCRL?+`+`AOEUIDHTNS_`+`:QJKXBMWVZ`,
  'colemak.low'   	: String.raw`${"`"}1234567890-=`     +`qwfpgjluy;[]`+`arstdhneio'`+`zxcvbkm,./`,
  'colemak.upp'   	: String.raw     `~!@#$%^&*()_+`     +`QWFPGJLUY:{}`+`ARSTDHNEIO"`+`ZXCVBKM<>?`,
  'workman.low'   	: String.raw`${"`"}1234567890-=`     +`qdrwbjfup;[]`+`ashtgyneoi'`+`zxmcvkl,./`,
  'workman.upp'   	: String.raw     `~!@#$%^&*()_+`     +`QDRWBJFUPP{}`+`ASHTGYNEOI"`+`ZXMCVKL<>?`,
  'asset.low'     	: String.raw`${"`"}1234567890-=`     +`qwjfgypul;[]`+`asetdhnior'`+`zxcvbkm,./`,
  'asset.upp'     	: String.raw     `~!@#$%^&*()_+`     +`QWJFGYPUL;{}`+`ASETDHNIO:"`+`ZXCVBKM<>?`,
  'colemak_dh.low'	: String.raw`${"`"}1234567890-=`     +`qwfpbjluy;[]`+`arstgkneio'`+`zxcdvmh,./`,
  'colemak_dh.upp'	: String.raw     `~!@#$%^&*()_+`     +`QWFPBJLUY:{}`+`ARSTGKNEIO"`+`ZXCDVMH<>?`,
  'neo2.low'      	: String.raw     `^1234567890-${"`"}`+`xvlcwkhgfqß´`+`uiaeosnrtdy`+`üöäpzbm,.j`,
  'neo2.upp'      	: String.raw     `ˇ°§ℓ»«$€„“”—¸`     +`XVLCWKHGFQẞ˜`+`UIAEOSNRTDY`+`ÜÖÄPZBM–•J`,
};
export const layouts = {'qwerty':'q','dvorak':'d','colemak':'c','colemak_dh':'cdh','workman':'w','neo2':'n','asset':'a'};
const button_name	= [];
export const lyt 	= {}; // q:qwerty, qwerty:q
for (const [lytName, lytAbbrv] of Object.entries(layouts)) {
  layout_string[lytName] = layout_string[lytName+'.low'] + layout_string[lytName+'.upp'];
  button_name.push('btn_'+lytName);
  lyt[lytName] 	= lytAbbrv;
  lyt[lytAbbrv]	= lytName;
}
const btn_prefix_len	= "btn_".length;
const buttons       	= document.getElementsByClassName("btn")	; // get a list of buttons to style later
const class_clicked 	= 'layout_button_clicked'               	; // class that applies a colored border
const btn_default   	= 'btn_qwerty'                          	; // default layout before any button is clicked

export const Case = Object.freeze({
  l	: Symbol("lower"),
  U	: Symbol("upper"),
});
function isLowerCaseLyt(str,lyt) { return layout_string[lyt+'.low'].includes(str);}
function isUpperCaseLyt(str,lyt) { return layout_string[lyt+'.upp'].includes(str);}
export function getCaseLyt(str,lyt) {
  if       	(isLowerCaseLyt(str,lyt))	{ return Case.l;
  } else if	(isUpperCaseLyt(str,lyt))	{ return Case.U;
  } else   	                         	{ return null;}
}
export function convertCaseLyt(char, lytNm='qwerty', CaseTo=null) {
  let caseCurrent	= getCaseLyt(char, lytNm);
  if (!caseCurrent) {return;}
  let lyt_from, lyt_to;
  switch (CaseTo) {
    case Case.U: // to upper
      lyt_from	= layout_string[lytNm+'.low'] ;
      lyt_to  	= layout_string[lytNm+'.upp'] ;
      break;
    case Case.l: // to lower
      lyt_from	= layout_string[lytNm+'.upp'] ;
      lyt_to  	= layout_string[lytNm+'.low'] ;
      break;
    default: // swap case
      switch (caseCurrent) {
        case Case.U: // to lower
          lyt_from	= layout_string[lytNm+'.upp'] ;
          lyt_to  	= layout_string[lytNm+'.low'] ;
          break;
        case Case.l: // to upper
          lyt_from	= layout_string[lytNm+'.low'] ;
          lyt_to  	= layout_string[lytNm+'.upp'] ;
          break;
        default:
          return;
      }
  }
  return lyt_to.charAt(lyt_from.indexOf(char)) || char;
}

export function convert(src, from, to) {
  let ret         	= '';
  const layoutFrom	= layout_string[from];
  const layoutTo  	= layout_string[to];
  for (var i = 0; i < src.length; i++) {
    ret += layoutTo.charAt(layoutFrom.indexOf(src.charAt(i))) || src.charAt(i);
  }
  return ret;
}
const changeElementLayoutTo = ((layoutTo) => { // change kbd elements (have layout attribute)
  gLyt.lbl = layoutTo;
  for (let el of document.querySelectorAll("*[lyt]")) {
    const layoutFrom	= el.getAttribute("lyt");
    const thisKey   	= el.textContent;
    if (layoutFrom === layoutTo) {
    } else {
      el.textContent	= convert(thisKey, lyt[layoutFrom], lyt[layoutTo]) || thisKey;
      el.setAttribute("lyt", layoutTo);
    }
  }
});
const changeKLELayoutTo = ((layoutFrom,layoutTo) => { // change labels of the KLE keymaps (no attribute)
  for (let el of document.querySelectorAll("div.keylabel.keylabel10.textsize2 > div")) {
    if (layoutFrom === layoutTo) {
    } else {
      const thisLbl	= el.textContent;
      let conv_l = (thisLbl === 'ẞ' ? 'ß' : thisLbl.toLowerCase());
      let conv   = convert(conv_l, lyt[layoutFrom], lyt[layoutTo]);
      let conv_u = (conv === 'ß' ? 'ẞ' : conv.toUpperCase());
      el.textContent	= conv_u || thisLbl;
    }
  }
});

button_name.forEach(((x) => { // x=button id, e.g., 'btn_qwerty'
  const btn = document.getElementById(x);
  if (btn) {
    btn.addEventListener("click", function() {
    const layoutTo = lyt[x.slice(btn_prefix_len)];	// 'btn_Qwerty' → Q → Qwerty
    changeElementLayoutTo(layoutTo);              	// change kbd buttons to ↑

    let btn_active = btn_default;
    for (var i = 0; i < button_name.length; i++) {
      let btn_i = document.getElementsByClassName("btn")[i];
      if (btn_i && btn_i.classList.contains(class_clicked)) {	// find last clicked button
        btn_active = btn_i.id;                               	// ...store its layout
        btn_i.classList.remove(class_clicked);               	// ...and remove its color border ()
      }
    }

    this.classList.add(class_clicked); // add color border to the clicked button

    const layoutForm = lyt[btn_active.slice(btn_prefix_len)];	// 'btn_Qwerty' → Q → Qwerty
    changeKLELayoutTo(layoutForm, layoutTo)
  }, false);}
}));
changeElementLayoutTo(gLyt.lbl);

File: src/layout_convert.py
import re

def getUserKeymap(keymap_len):
  # to use custom keymaps (non-QWERTY/non-Latin) in command mode
  # 1. add "lower"/"upper" keymaps below that match the position of QWERTY symbols
  # 2. Run command "NeoVintageous: Generate non-QWERTY key bindings"
  keymap = {
    #qwerty	      `12 34567890-=\ qwertyuiop[] asdfghjkl;' zxcvbnm,./
    "low"  	: R'''ё12 34567890-=\ йцукенгшщзхъ фывапролджэ ячсмитьбю.''',
    #qwerty	      ~!@ #$%^&*()_+| QWERTYUIOP{} ASDFGHJKL:" ZXCVBNM<>?
    "upp"  	: R'''Ё!@ #$%^&*()_+/ ЙЦУКЕНГШЩЗХЪ ФЫВАПРОЛДЖЭ ЯЧСМИТЬБЮ?''',
    # ↑ remap only unique non-Latin keys, leave symbols as is not to lose them in Latin layout
  }
  low = re.sub(r'\s','',keymap["low"])
  upp = re.sub(r'\s','',keymap["upp"])
  if not (ln := len(low)) == keymap_len:
    msg_error = f"keymap setting should have '{keymap_len}' characters, not '{ln}'"
    print(msg_error)
    return
  return({"low":low,"upp":upp}) # None to disable

class Symbol:
  def __init__(self, name=''):
    self.name = f"Symbol({name})"
  def __repr__(self):
    return self.name

from enum import Enum
class lyt(Enum):
  qwerty      = Symbol('qwerty')
  dvorak      = Symbol('dvorak')
  colemak     = Symbol('colemak')
  workman     = Symbol('workman')
  asset       = Symbol('asset')
  colemak_dh  = Symbol('colemak_dh')
  neo2        = Symbol('neo2')
  ru_pc       = Symbol('ru_pc')
  ru_mac      = Symbol('ru_mac')
  user        = Symbol('user')

class LayoutConverter:
  def __init__(self):
    self.isUser   = False
    layout_str        = {
      lyt.qwerty      :{'low' : R'''`1234567890-=\ qwertyuiop[] asdfghjkl;' zxcvbnm,./'''
       ,                'upp' : R'''~!@#$%^&*()_+| QWERTYUIOP{} ASDFGHJKL:" ZXCVBNM<>?'''},
      lyt.dvorak      :{'low' : R'''`1234567890[]\ ',.pyfgcrl/= aoeuidhtns- ;qjkxbmwvz'''
       ,                'upp' : R'''~!@#$%^&*(){}| "<>PYFGCRL?+ AOEUIDHTNS_ :QJKXBMWVZ'''},
      lyt.colemak     :{'low' : R'''`1234567890-=\ qwfpgjluy;[] arstdhneio' zxcvbkm,./'''
       ,                'upp' : R'''~!@#$%^&*()_+| QWFPGJLUY:{} ARSTDHNEIO" ZXCVBKM<>?'''},
      lyt.workman     :{'low' : R'''`1234567890-=\ qdrwbjfup;[] ashtgyneoi' zxmcvkl,./'''
       ,                'upp' : R'''~!@#$%^&*()_+| QDRWBJFUPP{} ASHTGYNEOI" ZXMCVKL<>?'''},
      lyt.asset       :{'low' : R'''`1234567890-=\ qwjfgypul;[] asetdhnior' zxcvbkm,./'''
       ,                'upp' : R'''~!@#$%^&*()_+| QWJFGYPUL;{} ASETDHNIO:" ZXCVBKM<>?'''},
      lyt.colemak_dh  :{'low' : R'''`1234567890-=\ qwfpbjluy;[] arstgkneio' zxcdvmh,./'''
       ,                'upp' : R'''~!@#$%^&*()_+| QWFPBJLUY:{} ARSTGKNEIO" ZXCDVMH<>?'''},
      lyt.neo2        :{'low' : R'''^1234567890-`\ xvlcwkhgfqß´ uiaeosnrtdy üöäpzbm,.j'''
       ,                'upp' : R'''ˇ°§ℓ»«$€„“”—¸| XVLCWKHGFQẞ˜ UIAEOSNRTDY ÜÖÄPZBM–•J'''},
      lyt.ru_pc       :{'low' : R'''ё1234567890-=\ йцукенгшщзхъ фывапролджэ ячсмитьбю.'''
       ,                'upp' : R'''Ё!"№;%:?*()_+/ ЙЦУКЕНГШЩЗХЪ ФЫВАПРОЛДЖЭ ЯЧСМИТЬБЮ,'''},
      lyt.ru_mac      :{'low' : R''']1234567890-=ё йцукенгшщзхъ фывапролджэ ячсмитьбю/'''
       ,                'upp' : R'''[!"№;%,.;()_+Ё ЙЦУКЕНГШЩЗХЪ ФЫВАПРОЛДЖЭ ЯЧСМИТЬБЮ?'''},
    }
    for k,v in layout_str.items():
      v['low'] = v['low'].replace(' ','')
      v['upp'] = v['upp'].replace(' ','')
    keymap_len  = len(layout_str[lyt.qwerty]['low'])

    userKeymap = getUserKeymap(keymap_len)
    if userKeymap:
      layout_str[lyt.user] = userKeymap
      self.isUser = True

    translations = dict() # generate translation dictionaries for use in str.translate(dict)
    for   layout_from in layout_str:
      translations[layout_from] = dict()
      for layout_to   in layout_str:
        string_from = layout_str[layout_from  ]['low'] + layout_str[layout_from ]['upp']
        string_to   = layout_str[layout_to    ]['low'] + layout_str[layout_to   ]['upp']
        translations[layout_from][layout_to] =  str.maketrans(string_from, string_to)
    self.layout_str   = layout_str
    self.layouts      = list(layout_str.keys())
    self.translations = translations

  def convert(self, src, layout_from, layout_to):
    translations  = self.translations
    layouts       = self.layouts
    if not layout_from in layouts: # todo: or fail silently and return src?
      msg_error = f"'{layout_from}' invalid, must be one of '{layouts}')"
      print(msg_error)
      return None
    if not layout_to   in layouts:
      msg_error = f"'{layout_to}' invalid, must be one of '{layouts}')"
      print(msg_error)
      return None
    return src.translate(translations[layout_from][layout_to])
